pkg/agent/monitortool/latency_store.go

```diff
@@ -0,0 +1,248 @@
+
+// updateNodeMap updates the nodeTargetIPsMap with the IPs of the given Node.
+func (s *LatencyStore) updateNodeMap(node *corev1.Node) {
+	nodeIPs, err := s.getNodeIPs(node)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get IPs for Node", "nodeName", node.Name)
+		return
+	}
+
+	s.nodeTargetIPsMap[node.Name] = nodeIPs
+}
+
+// getNodeIPs returns the target IPs of the given Node based on the agent mode.
+func (s *LatencyStore) getNodeIPs(node *corev1.Node) ([]net.IP, error) {
+	if s.isNetworkPolicyOnly {
+		transportIPs, err := getTransportIPs(node)
+		if err != nil {
+			return nil, err
+		}
+
+		return transportIPs, nil
+	} else {
+		gw0IPs, err := getGWIPs(node)
+		if err != nil {
+			return nil, err
+		}
+
+		return gw0IPs, nil
+	}
+}
+
+// getTransportIPs returns the transport IPs of the given Node.
+func getTransportIPs(node *corev1.Node) ([]net.IP, error) {
+	var transportIPs []net.IP
+	ips, err := k8s.GetNodeTransportAddrs(node)
+	if err != nil {
+		return transportIPs, err
+	}
+
+	if ips.IPv4 != nil {
+		transportIPs = append(transportIPs, ips.IPv4)
+	}
+	if ips.IPv6 != nil {
+		transportIPs = append(transportIPs, ips.IPv6)
+	}
+
+	return transportIPs, nil
+}
+
+// getGWIPs returns the gateway IPs of the given Node.
+func getGWIPs(node *corev1.Node) ([]net.IP, error) {
+	var gwIPs []net.IP
+
+	podCIDRStrs := getPodCIDRsOnNode(node)
+	if len(podCIDRStrs) == 0 {
+		return nil, errors.New("node does not have a PodCIDR")
+	}
+
+	for _, podCIDR := range podCIDRStrs {
+		peerPodCIDRAddr, _, err := net.ParseCIDR(podCIDR)
+		if err != nil {
+			return nil, err
+		}
+
+		// Add first IP in CIDR to the map
+		peerGatewayIP := ip.NextIP(peerPodCIDRAddr)
+		gwIPs = append(gwIPs, peerGatewayIP)
+	}
+
+	return gwIPs, nil
+}
+
+// getPodCIDRsOnNode returns the PodCIDRs of the given Node.
+func getPodCIDRsOnNode(node *corev1.Node) []string {
+	if node.Spec.PodCIDRs != nil {
+		return node.Spec.PodCIDRs
+	}
+
+	if node.Spec.PodCIDR == "" {
+		return nil
+	}
+	return []string{node.Spec.PodCIDR}
+}
+
+// ListNodeIPs returns the list of all Node IPs in the latency store.
+func (s *LatencyStore) ListNodeIPs() []net.IP {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	// Allocate a slice with a capacity equal to twice the size of the map,
+	// as we can have up to 2 IP addresses per Node in dual-stack case.
+	nodeIPs := make([]net.IP, 0, 2*len(s.nodeTargetIPsMap))
+	for _, ips := range s.nodeTargetIPsMap {
+		nodeIPs = append(nodeIPs, ips...)
+	}
+
+	return nodeIPs
+}
+
+// DeleteStaleNodeIPs deletes the stale Node IP from the nodeIPLatencyMap.
+func (s *LatencyStore) DeleteStaleNodeIPs() {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	// We will try to delete the stale Node IP from the nodeIPLatencyMap
+	// If a PodCIDR is used by more than one Node in a short time and the Node is deleted,
+	// the recv packet from the Node will update the latency of the Node.
+	for nodeIP := range s.nodeIPLatencyMap {
+		if _, ok := s.nodeTargetIPsMap[nodeIP]; !ok {
```

Comment:
This is wrong, `nodeTargetIPsMap` is indexed by Node name, not Node IP!

The code should:
- build a set (https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#Set) of active Node IPs by iterating over all the *values* (not keys) in `s.nodeTargetIPsMap`
- iterate over the `s.nodeIPLatencyMap` keys and delete all of those who are not in the set we have built

Reply:
Please also add a unit test for this function as it is straightforward to do (unlike unit tests for the monitor, which we are postponing to a subsequent PR)

Reply:
Oh sorry for the problem due to my negligence, I have fixed the problem.

Reply:
I don't see any update to this code

Reply:
The commit is submitted just now.